### PR TITLE
MessageContext inherits from IExtendable

### PR DESCRIPTION
**Pull request by danielmarbach**
Aligning context with what has been proposed by @andreasohlund in #4468 

According to https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md this is a non-breaking change

> Adding an interface implementation to a type
> This is acceptable because it will not adversely affect existing clients. Any changes which could be made to the type being changed in this situation, will have to work within the boundaries of acceptable changes defined here, in order for the new implementation to remain acceptable. Extreme caution is urged when adding interfaces that directly affect the ability of the designer or serializer to generate code or data, that cannot be consumed down-level. An example is the ISerializable interface.

### CODE REVIEW

File: src/NServiceBus.AcceptanceTests/Satellites/When_a_message_is_available.cs [hunk 1]
```diff
@@ -56,7 +56,7 @@ protected override void Setup(FeatureConfigurationContext context)
                         {
                             var testContext = builder.Build<Context>();
                             testContext.MessageReceived = true;
-                            testContext.TransportTransactionAddedToContext = ReferenceEquals(pushContext.Context.Get<TransportTransaction>(), pushContext.TransportTransaction);
+                            testContext.TransportTransactionAddedToContext = ReferenceEquals(pushContext.Extensions.Get<TransportTransaction>(), pushContext.TransportTransaction);
```
Comment (bording) [inline src/NServiceBus.AcceptanceTests/Satellites/When_a_message_is_available.cs]: Not sure it makes sense to change in this PR, but we need to go and scrub the remaining "pushContext" usages in the code.

File: src/NServiceBus.Core/Transports/MessageContext.cs [hunk 2]
```diff
@@ -68,6 +68,12 @@ public MessageContext(string messageId, Dictionary<string, string> headers, byte
         /// <summary>
         /// Context provided by the transport.
         /// </summary>
-        public ContextBag Context { get; }
+        [ObsoleteEx(Message = "Usage of context is discouraged.", ReplacementTypeOrMember = "Extensions", RemoveInVersion = "7")]
```
Comment (bording) [inline src/NServiceBus.Core/Transports/MessageContext.cs]: How about "Usage of 'Context' is discouraged." instead? 

Alternatively, does that message really gain us anything over just having the obsolete there anyway?

File: src/NServiceBus.Core/Transports/MessageContext.cs [hunk 3]
```diff
@@ -34,7 +34,7 @@ public MessageContext(string messageId, Dictionary<string, string> headers, byte
             Headers = headers;
             Body = body;
             MessageId = messageId;
-            Context = context;
+            Extensions = context;
```
Comment (bording) [inline src/NServiceBus.Core/Transports/MessageContext.cs]: What about the parameter name, should it be `extensions` too?

File: src/NServiceBus.AcceptanceTests/Satellites/When_a_message_is_available.cs (hunk 1 above)
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Satellites/When_a_message_is_available.cs]: You mean renaming it to message context?

**Comment by danielmarbach** [comment]
@bording addressed your comments

File: src/NServiceBus.AcceptanceTests/Satellites/When_a_message_is_available.cs (hunk 1 above)
Comment (bording) [inline src/NServiceBus.AcceptanceTests/Satellites/When_a_message_is_available.cs]: Yeah, that's what I meant. `PushContext` was the old name of the class, so there were variables that weren't renamed.

File: src/NServiceBus.Core/Transports/MessageContext.cs [hunk 4]
```diff
@@ -21,20 +21,20 @@ public class MessageContext
         /// It also allows the transport to communicate to the pipeline to abort if possible. Transports should check if the token
         /// has been aborted after invoking the pipeline and roll back the message accordingly.
         /// </param>
-        /// <param name="context">Any context that the transport wants to be available on the pipeline.</param>
-        public MessageContext(string messageId, Dictionary<string, string> headers, byte[] body, TransportTransaction transportTransaction, CancellationTokenSource receiveCancellationTokenSource, ContextBag context)
+        /// <param name="extensions">Any extensions that the transport wants to be available on the pipeline.</param>
```
Comment (bording) [inline src/NServiceBus.Core/Transports/MessageContext.cs]: The comment here should probably be revised to something like "Any additional values" or "Any additional properties".

"Any extensions" doesn't really make sense.

Comment (bording) [inline src/NServiceBus.Core/Transports/MessageContext.cs]: Perhaps "A <see cref="ContextBag" /> which can be used to extend the current object." like used in #4468

**Comment by danielmarbach** [comment]
@bording pushed an update

File: src/NServiceBus.sln.DotSettings [hunk 5]
```diff
@@ -596,6 +596,7 @@ II.2.12 &lt;HandlesEvent /&gt;&#xD;
 	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002EJavaScript_002ECodeStyle_002ESettingsUpgrade_002EJsCodeFormatterSettingsUpgrader/@EntryIndexedValue">True</s:Boolean>
 	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002EJavaScript_002ECodeStyle_002ESettingsUpgrade_002EJsParsFormattingSettingsUpgrader/@EntryIndexedValue">True</s:Boolean>
 	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002EJavaScript_002ECodeStyle_002ESettingsUpgrade_002EJsWrapperSettingsUpgrader/@EntryIndexedValue">True</s:Boolean>
+	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002EXml_002ECodeStyle_002EFormatSettingsUpgrade_002EXmlMoveToCommonFormatterSettingsUpgrade/@EntryIndexedValue">True</s:Boolean>
```
Comment (bording) [inline src/NServiceBus.sln.DotSettings]: What does this do? Is this from RepoStandards?

File: src/NServiceBus.Core/Transports/MessageContext.cs (hunk 4 above)
Comment (bording) [inline src/NServiceBus.Core/Transports/MessageContext.cs]: @danielmarbach You said you pushed an update, but I don't see any change to this parameter comment?

**Comment by danielmarbach** [comment]
Must be a client glitch then I'll fix. I have it local

File: src/NServiceBus.sln.DotSettings (hunk 5 above)
Comment (danielmarbach) [inline src/NServiceBus.sln.DotSettings]: From rider. I think I accidentally staged this instead of the param change

**Comment by timbussmann** [comment]
@danielmarbach ping

**Comment by danielmarbach** [comment]
@timbussmann pong ;)

@bording done

File: src/NServiceBus.Core.Tests/Sagas/SagaMetadataCreationTests.cs [hunk 6]
```diff
@@ -222,6 +222,14 @@ public void GetEntityClrTypeFromInheritanceChain()
             Assert.AreEqual(typeof(SagaWithInheritanceChain.SagaData), metadata.SagaEntityType);
         }
 
+        [Test]
+        public void Foo()
```
Comment (bording) [inline src/NServiceBus.Core.Tests/Sagas/SagaMetadataCreationTests.cs]: @danielmarbach Looks like some more unrelated stuff has snuck into this PR!

**Comment by danielmarbach** [comment]
I think I'm not capable of switching the git client which I've just proven
with this PR. What a mess, sorry. Can you reset this file, force push and
then merge it or do you really want to wait another 48h?

Brandon Ording <notifications@github.com> schrieb am Mo. 6. März 2017 um
23:17:

**Comment by bording** [comment]
> I think I'm not capable of switching the git client which I've just proven
with this PR. What a mess, sorry. Can you reset this file, force push and
then merge it or do you really want to wait another 48h?

@danielmarbach I'll take care of it and merge it.

Let me know if you need some git training 😉
